src/VisualStudio/Core/Test/ProjectSystemShim/VisualStudioProjectTests/AnalyzerReferenceTests.vb

```diff
@@ -99,84 +122,195 @@ Namespace Microsoft.VisualStudio.LanguageServices.UnitTests.ProjectSystemShim
+                        Path.Combine(vsixDir, "Util.dll")
+                    }, GetAnalyzerFilePaths(getAnalyzerRefs()).OrderBy(Function(x) Path.GetFileName(x), StringComparer.Ordinal))
+
+                project.RemoveAnalyzerReference(Path.Combine(otherDir, "SpellChecker.dll"))
+                Assert.Equal(2, getAnalyzerRefs().Count)
+                project.RemoveAnalyzerReference(Path.Combine(vsixDir, "Util.dll"))
+                Assert.Equal(2, getAnalyzerRefs().Count)
+                project.RemoveAnalyzerReference(Path.Combine(vsixDir, ProjectSystemProject.RazorSourceGenaratorDllName))
+                Assert.Empty(getAnalyzerRefs())
+            End Using
+        End Function
+
+        <WpfFact>
+        Public Async Function RazorSourceGenerator_FromVsix_ErrorCases() As Task
+            Using environment = New TestEnvironment()
+
+                Dim vsixDir = CreateAnalyzerDir(ProjectSystemProject.RazorSourceGenaratorDllName, "Analyzer.dll")
+
+                Dim providerFactory = DirectCast(environment.ExportProvider.GetExportedValue(Of IVisualStudioDiagnosticAnalyzerProviderFactory), MockVisualStudioDiagnosticAnalyzerProviderFactory)
+                providerFactory.Extensions =
                 {
-                    Path.Combine(TempRoot.Root, "Dir", "File.dll")
-                }, environment.Workspace.CurrentSolution.Projects.Single().AnalyzerReferences.Select(Function(r) r.FullPath))
+                    ({
+                        Path.Combine(vsixDir, ProjectSystemProject.RazorSourceGenaratorDllName),
+                        Path.Combine(vsixDir, "Util.dll")
+                     },
+                     ProjectSystemProject.RazorVsixExtensionId)
+                }
+
+                Dim project = Await environment.ProjectFactory.CreateAndAddToWorkspaceAsync(
+                    "Project", LanguageNames.CSharp, CancellationToken.None)
+
+                Dim getAnalyzerRefs =
+                    Function() As IReadOnlyList(Of AnalyzerReference)
+                        Return environment.Workspace.CurrentSolution.Projects.Single().AnalyzerReferences
+                    End Function
+
+                Dim sdkDir = CreateAnalyzerDir(ProjectSystemProject.RazorSourceGenaratorDllName, "Util.dll")
+
+                project.AddAnalyzerReference(Path.Combine(sdkDir, ProjectSystemProject.RazorSourceGenaratorDllName))
+                project.AddAnalyzerReference(Path.Combine(sdkDir, "Util.dll"))
+
+                project.RemoveAnalyzerReference(Path.Combine(sdkDir, "Util.dll"))
+
+                ' This does not throw because within the razor generator only the actual generator dll triggers real 
+                ' add / removes. At a glance this may seem problematic but it's an unavoidable outcome once the 
+                ' project system begins mapping / mangling analyzer paths. There will be cases where existing names 
+                ' do not map to analyzers and it ends up with this behavior. 
```

Comment:
This could be fixed by putting a bit more logic into `ProjectSystemProject` where by we keep track of the original paths that were mapped to empty. Then we could have the Add / Remove calls manipulate that table and get back throwing behavior for double remove. Ended up deciding against that because it's a half measure. Yes we get back the Add / Remove semantics but it doesn't fix the fact that `AnalyzerReferences` in the project don't change. 

I could be convinced to go the other direction.

Reply:
I'm fine with this, since at some point I hope this gets properly refactored. :smile: